### PR TITLE
Burn in flow_run_id, remote_url, sha into package meta data

### PR DESCRIPTION
**Pull request by dbast**
This is ready for discussion, if that concept would be excepted or what changes to apply to it: It is based on the idea to make it possible to trace back existing packages to their origin (repo, commit and CI run) for audit purposes, but also for easier rebuilds in the future of older packages by knowing from what repo and commit they have been created.

This uses the conda-build 3.21.8 feature to add extra meta data to the resulting package to include the information, from which git-repo/feedstock, commit sha1 and CI RUN ID a package was created.

How can I test that? Just using an existing feedstock, where I have access and do a local rerender with rotating through all the different CI providers?

@wolfv Can I assume this also works with mamba-build/boa as it uses conda-build in the end to create the package?

Thanks!

cc @cjmartian, who implemented that in conda-build.

Docs:
* https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables
* https://docs.drone.io/pipeline/environment/reference/
* https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
* https://docs.travis-ci.com/user/environment-variables/
* https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables
* https://www.appveyor.com/docs/environment-variables/

Checklist
* [x] Added a ``news`` entry

### CODE REVIEW

**Comment by wolfv** [comment]
We should test that mambabuild forwards the arguments properly but in general this should work, yeah!

**Comment by cjmartian** [comment]
This all looks correct to me, haven't really touched conda forge CI before though.

**Comment by h-vetinari** [comment]
This sounds like a cool addition! Any updates on this?

**Comment by jaimergp** [comment]
@dbast - is this is still [WIP]?

**Comment by dbast** [comment]
@jaimergp Thanks for looking into this .. the PR is not really WIP anymore, but hard to test:
* the used conda-build feature (with tests) to burn in that meta data was added via this PR https://github.com/conda/conda-build/pull/4303 
* The `write_about_json(_)` function, which in the end burns in that data in conda-build, unfortunately does not write anything to the build log, which would help to verify that the right extra meta data is burned into the package.
* Thus this can be only verified by inspecting created packages, which is harder than just looking at the build log.
* So even if this current conda-smithy PR would be correct atm, it would be very hard over time to warrant that any future changes (like newly added CI providers) don't degrade this functionality in conda-smithy.
* So best thing would be to add a log line in conda-build that logs the extra data... after that we can see in every build log, if we consistently burn-in the right meta data... and even helps to co-relate the build log with the package, by both showing the same hashes.

**Comment by dbast** [comment]
conda-build 3.26.0 is available and makes testing this much easier by logging the extra-meta data.

**Comment by jakirkham** [comment]
Thanks Daniel! 🙏

Think something like this is a good idea

There are other use cases we might want to consider. For example adding links on Anaconda.org for the recipe source, etc. For example ( https://github.com/conda/conda-build/pull/2489 )? What tool we use for this? Whether we package build logs?

This may benefit from a design discussion (and maybe a CEP) to ensure broad support for this kind of functionality and in a way that is generally useful

**Comment by isuruf** [comment]
@conda-forge/core, this is ready for a review

**Comment by isuruf** [comment]
@jakirkham, I changed the variable names to the ones used by Anaconda.

Here's the metadata from the pkgs/main/python pkg
```yaml
extra:
  copy_test_source_files: true
  feedstock-name: python
  final: true
  flow_run_id: 5a3cdab7-40d3-44ab-8d4e-16946a951bef
  recipe-maintainers:
    - isuruf
    - jakirkham
    - katietz
    - mbargull
    - mingwandroid
    - msarahan
    - ocefpaf
    - pelson
    - scopatz
    - xhochy
  remote_url: git@github.com:AnacondaRecipes/python-feedstock.git
  sha: 4ad3a1da3343d359f1961b7a0cb721cdd396f6b9
```

File: conda_smithy/templates/azure-pipelines-linux.yml.tmpl [hunk 1]
```diff
@@ -81,7 +84,7 @@ jobs:
 {%- if azure.store_build_artifacts %}
   - script: |
         export CI=azure
-        export CI_RUN_ID=$(build.BuildNumber).$(system.JobAttempt)
+        export flow_run_id=$(build.BuildNumber).$(system.JobAttempt)
```
Comment (jaimergp) [inline conda_smithy/templates/azure-pipelines-linux.yml.tmpl:87]:
```suggestion
        export flow_run_id=azure_$(build.BuildNumber).$(system.JobAttempt)
```

Should this be prefixed too?

File: conda_smithy/templates/azure-pipelines-osx.yml.tmpl [hunk 2]
```diff
@@ -34,7 +37,7 @@ jobs:
 {%- if azure.store_build_artifacts %}
   - script: |
       export CI=azure
-      export CI_RUN_ID=$(build.BuildNumber).$(system.JobAttempt)
+      export flow_run_id=$(build.BuildNumber).$(system.JobAttempt)
```
Comment (jaimergp) [inline conda_smithy/templates/azure-pipelines-osx.yml.tmpl:40]:
```suggestion
      export flow_run_id=azure_$(build.BuildNumber).$(system.JobAttempt)
```

same

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 3]
```diff
@@ -40,6 +40,9 @@ jobs:
         PYTHONUNBUFFERED: 1
         CONFIG: $(CONFIG)
         CI: azure
+        flow_run_id: $(Build.BuildNumber).$(System.JobAttempt)
```
Comment (jaimergp) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:43]:
```suggestion
        flow_run_id: azure_$(Build.BuildNumber).$(System.JobAttempt)
```

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 4]
```diff
@@ -52,7 +55,7 @@ jobs:
 {%- if azure.store_build_artifacts %}
     - script: |
         set CI=azure
-        set CI_RUN_ID=$(build.BuildNumber).$(system.JobAttempt)
+        set flow_run_id=$(build.BuildNumber).$(system.JobAttempt)
```
Comment (jaimergp) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:58]:
```suggestion
        set flow_run_id=azure_$(build.BuildNumber).$(system.JobAttempt)
```

File: news/burn_in_ids.rst [hunk 5]
```diff
@@ -0,0 +1,26 @@
+**Added:**
+
+* Add flow_run_id (CI provider specific), remote_url and sha as extra-meta data to packages.
+  Enables tracing back packages to a specific commit in a feedstocks and to a specifc CI run.
+  When packages are built using build-locally.py only GIT_SHA1 will have a non-empty value.
+  Requires conda-build >=3.21.8.
```
Comment (jaimergp) [inline news/burn_in_ids.rst:6]:
```suggestion
  Requires conda-build >=3.21.8. (#1577)
```

Comment (jaimergp) [inline news/burn_in_ids.rst:6]:
```suggestion
* Add `flow_run_id` (CI provider specific), `remote_url` and `sha` as extra-meta data to packages.
  Enables tracing back packages to a specific commit in a feedstock and to a specific CI run.
  When packages are built using `build-locally.py` only `sha` will have a non-empty value.
  Requires `conda-build >=3.21.8`. (#1577)
```

File: conda_smithy/templates/create_conda_build_artifacts.sh.tmpl
Comment (jaimergp) [inline conda_smithy/templates/create_conda_build_artifacts.sh.tmpl:1]: Do we need to revert the changes here now?

Comment (isuruf) [reply conda_smithy/templates/create_conda_build_artifacts.sh.tmpl:1]: Done

**Comment by dbast** [comment]
Yay, thanks the help here!

**Review by xhochy — APPROVED** [review]

**Comment by jaimergp** [comment]
Woohoo! We have provenance data now! Look at this https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fnoarch%2Fmakim-1.8.3-pyh707e725_0.conda

image


"Provenance" cell in the table.

**Comment by AndresGuzman-Ballen** [comment]
This is absolutely amazing indeed! Out of curiosity, does this mean that only packages built starting today will include the remote_url metadata? Or is this going to be retroactively applied and in a couple of weeks/months when CI has made its way to all the packages, then theoretically any conda package that has ever existed in conda-forge channel will have a remote_url entry?

**Comment by jaimergp** [comment]
I'm afraid this PR only covers new artifacts. I'm not even sure how we could retroactively find the needed bits.

**Comment by AndresGuzman-Ballen** [comment]
I had a feeling. If you've ever watched the movie "Primer", the same time-travel logic applies where you can only travel back to the time the time-machine was invented LOL

**Comment by jakirkham** [comment]
Or it could be like the game [Continuum]( https://en.wikipedia.org/wiki/Continuum_(role-playing_game) ). You can change the past, but you have to change it back when you are done 😉
